core/iwasm/compilation/aot_emit_control.c

```diff
@@ -1325,22 +1325,163 @@ aot_handle_next_reachable_block(AOTCompContext *comp_ctx,
+                POP(value, block_dst->param_types[param_index]);
+                ADD_TO_PARAM_PHIS(block_dst, value, param_index);
+                values[param_index] = value;
+            }
+            for (i = 0; i < block_dst->param_count; i++) {
+                PUSH(values[i], block_dst->param_types[i]);
+            }
+            wasm_runtime_free(values);
+            values = NULL;
+        }
+
+        BUILD_COND_BR(value_cmp, block_dst->llvm_entry_block, llvm_else_block);
+
+        /* Move builder to else block */
+        SET_BUILDER_POS(llvm_else_block);
+    }
+    else {
+        /* Dest block is Block/If/Function block */
+        /* Create the end block */
+        if (!block_dst->llvm_end_block) {
+            format_block_name(name, sizeof(name), block_dst->block_index,
+                              block_dst->label_type, LABEL_END);
+            CREATE_BLOCK(block_dst->llvm_end_block, name);
+            if ((next_llvm_end_block = find_next_llvm_end_block(block_dst)))
+                MOVE_BLOCK_BEFORE(block_dst->llvm_end_block,
+                                  next_llvm_end_block);
+        }
+
+        /* Set reachable flag and create condition br IR */
+        block_dst->is_reachable = true;
+
+        /* Handle result values */
+        if (block_dst->result_count) {
+            size = sizeof(LLVMValueRef) * (uint64)block_dst->result_count;
+            if (size >= UINT32_MAX
+                || !(values = wasm_runtime_malloc((uint32)size))) {
+                aot_set_last_error("allocate memory failed.");
+                goto fail;
+            }
+            CREATE_RESULT_VALUE_PHIS(block_dst);
+            for (i = 0; i < block_dst->result_count; i++) {
+                result_index = block_dst->result_count - 1 - i;
+                POP(value, block_dst->result_types[result_index]);
+                values[result_index] = value;
+                ADD_TO_RESULT_PHIS(block_dst, value, result_index);
+            }
+            for (i = 0; i < block_dst->result_count; i++) {
+                PUSH(values[i], block_dst->result_types[i]);
+            }
+            wasm_runtime_free(values);
+            values = NULL;
+        }
+
+        /* Condition jump to end block */
+        BUILD_COND_BR(value_cmp, block_dst->llvm_end_block, llvm_else_block);
+
+        /* Move builder to else block */
+        SET_BUILDER_POS(llvm_else_block);
+    }
+
+    return true;
+fail:
+    if (values)
+        wasm_runtime_free(values);
+    return false;
+}
+
 bool
 aot_compile_op_br_on_null(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
                           uint32 br_depth, uint8 **p_frame_ip)
 {
     LLVMValueRef gc_obj, value_cmp;
 
-    GET_GC_REF_FROM_STACK(gc_obj);
+    if (!commit_gc_and_check_suspend_flags(comp_ctx, func_ctx, br_depth,
+                                           p_frame_ip)) {
+        return false;
+    }
+
+    POP_GC_REF(gc_obj);
```

Comment:
Why we must pop out the gc_obj here?

Reply:
It is to meet the requirements of `compile_gc_cond_br` below, which may take the operands from the stack to the target block (popped from stacks, ADD_TO_PARAM_PHIS, and then pushed to stacks again). Since the operands passed to the target block don't include this gc_obj (it is null when br), we pop it and then push back after `compile_gc_cond_br`.
See also:
https://github.com/WebAssembly/function-references/blob/master/proposals/function-references/Overview.md#optional-references

Reply:
Got it, thanks!